crates/noirc_frontend/src/hir_def/types.rs

```diff
@@ -0,0 +1,273 @@
+}
+
+impl Type {
+    // These are here so that the code is more readable.
+    // Type::WITNESS vs Type::FieldElement(FieldElementType::Private)
+    pub const WITNESS: Type = Type::FieldElement(FieldElementType::Private);
+    pub const CONSTANT: Type = Type::FieldElement(FieldElementType::Constant);
+    pub const PUBLIC: Type = Type::FieldElement(FieldElementType::Public);
+}
+
+impl From<&Type> for AbiType {
+    fn from(ty: &Type) -> AbiType {
+        ty.as_abi_type()
+    }
+}
+
+impl std::fmt::Display for Type {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Type::FieldElement(fe_type) => write!(f, "{} Field", fe_type),
+            Type::Array(fe_type, size, typ) => write!(f, "{} {}{}", fe_type, size, typ),
+            Type::Integer(fe_type, sign, num_bits) => match sign {
+                Signedness::Signed => write!(f, "{} i{}", fe_type, num_bits),
+                Signedness::Unsigned => write!(f, "{} u{}", fe_type, num_bits),
+            },
+            Type::Struct(vis, s) => write!(f, "{} {}", vis, s.borrow()),
+            Type::Bool => write!(f, "bool"),
+            Type::Unit => write!(f, "()"),
+            Type::Error => write!(f, "error"),
+            Type::Unspecified => write!(f, "unspecified"),
+        }
+    }
+}
+
+impl Type {
+    // Returns true if the Type can be used in a Private statement
+    pub fn can_be_used_in_priv(&self) -> bool {
+        match self {
+            Type::FieldElement(FieldElementType::Private) => true,
+            Type::Integer(field_type, _, _) => field_type == &FieldElementType::Private,
+            Type::Error => true,
+            _ => false,
+        }
+    }
+
+    // A feature of the language is that `Field` is like an
+    // `Any` type which allows you to pass in any type which
+    // is fundamentally a field element. E.g all integer types
+    pub fn is_super_type_of(&self, argument: &Type) -> bool {
+        // Avoid reporting duplicate errors
+        if self == &Type::Error || argument == &Type::Error {
+            return true;
+        }
+
+        // if `self` is a `Field` then it is a super type
+        // if the argument is a field element
+        if let Type::FieldElement(FieldElementType::Private) = self {
+            return argument.is_field_element();
+        }
+
+        // For composite types, we need to check they are structurally the same
+        // and then check that their base types are super types
+        if let (Type::Array(_, param_size, param_type), Type::Array(_, arg_size, arg_type)) =
+            (self, argument)
+        {
+            let is_super_type = param_type.is_super_type_of(arg_type);
+            let arity_check = param_size.is_a_super_type_of(arg_size);
+            return is_super_type && arity_check;
+        }
+
+        // XXX: Should we also allow functions that ask for u16
```

Comment:
I think we should have the user explicitly do this with a cast. So the only special behaviour is on the `Field` type

Reply:
This is an existing note from ast/mod.rs since the old `Type` in that file was moved here (this new types.rs file is largely unchanged). We could address this in a separate PR though

Reply:
Sounds good